### ISSUE TITLE
bpf_timer is defined multiple times in build/redbpf-probes-xxx/out/gen_bindings.rs file

### ISSUE THREAD

**Issue by Sherlock-Holo** [foniod/redbpf#218]
I use redbpf 2.1 and cargo-bpf(with llvm13, command-line features), then I build my old project, but it reports

> ```
> error[E0428]: the name `bpf_timer` is defined multiple times
>       --> /home/sherlock/git/learn_bpf/target/bpf/release/build/redbpf-probes-22900dca2d1af05a/out/gen_bindings.rs:240875:5
>        |
> 139094 |     pub struct bpf_timer {
>        |     -------------------- previous definition of the type `bpf_timer` here
> ...
> 240875 |     pub struct bpf_timer {
>        |     ^^^^^^^^^^^^^^^^^^^^ `bpf_timer` redefined here
>        |
>        = note: `bpf_timer` must be defined only once in the type namespace of this module
> 
> error[E0119]: conflicting implementations of trait `core::clone::Clone` for type `bindings::gen_bindings::bpf_timer`
>       --> /home/sherlock/git/learn_bpf/target/bpf/release/build/redbpf-probes-22900dca2d1af05a/out/gen_bindings.rs:240874:27
>        |
> 139093 |     #[derive(Debug, Copy, Clone)]
>        |                           ----- first implementation here
> ...
> 240874 |     #[derive(Debug, Copy, Clone)]
>        |                           ^^^^^ conflicting implementation for `bindings::gen_bindings::bpf_timer`
>        |
>        = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
> 
> error[E0119]: conflicting implementations of trait `core::marker::Copy` for type `bindings::gen_bindings::bpf_timer`
>       --> /home/sherlock/git/learn_bpf/target/bpf/release/build/redbpf-probes-22900dca2d1af05a/out/gen_bindings.rs:240874:21
>        |
> 139093 |     #[derive(Debug, Copy, Clone)]
>        |                     ---- first implementation here
> ...
> 240874 |     #[derive(Debug, Copy, Clone)]
>        |                     ^^^^ conflicting implementation for `bindings::gen_bindings::bpf_timer`
>        |
>        = note: this error originates in the derive macro `Copy` (in Nightly builds, run with -Z macro-backtrace for more info)
> 
> error[E0119]: conflicting implementations of trait `core::fmt::Debug` for type `bindings::gen_bindings::bpf_timer`
>       --> /home/sherlock/git/learn_bpf/target/bpf/release/build/redbpf-probes-22900dca2d1af05a/out/gen_bindings.rs:240874:14
>        |
> 139093 |     #[derive(Debug, Copy, Clone)]
>        |              ----- first implementation here
> ...
> 240874 |     #[derive(Debug, Copy, Clone)]
>        |              ^^^^^ conflicting implementation for `bindings::gen_bindings::bpf_timer`
>        |
>        = note: this error originates in the derive macro `Debug` (in Nightly builds, run with -Z macro-backtrace for more info)
> ```

**Comment by Sherlock-Holo** [comment]
if I specify `KERNEL_SOURCE=/usr/lib/modules/5.15.2-arch1-1` env, it build success, but why it can't build without the env?

**Comment by rhdxmr** [comment]
@Sherlock-Holo 
I wonder your build environment. What distro do you use?
And do you compile your project in a docker container?

**Comment by Sherlock-Holo** [comment]
I am still using archlinux with llvm13

**Comment by Sherlock-Holo** [comment]
> uname -a
> Linux sherlockholo-pc 5.15.2-arch1-1 #1 SMP PREEMPT Fri, 12 Nov 2021 19:22:10 +0000 x86_64 GNU/Linux

> rustc --version --verbose
> rustc 1.58.0-nightly (e90c5fbbc 2021-11-12)
> binary: rustc
> commit-hash: e90c5fbbc5df5c81267747daeb937d4e955ce6ad
> commit-date: 2021-11-12
> host: x86_64-unknown-linux-gnu
> release: 1.58.0-nightly
> LLVM version: 13.0.0

> pacman -Qi llvm
> Name            : llvm
> Version         : 13.0.0-3
> Description     : Collection of modular and reusable compiler and toolchain technologies
> Architecture    : x86_64
> URL             : https://llvm.org/
> Licenses        : custom:Apache 2.0 with LLVM Exception
> Groups          : None
> Provides        : None
> Depends On      : llvm-libs  perl
> Optional Deps   : None
> Required By     : None
> Optional For    : clang
> Conflicts With  : None
> Replaces        : None
> Installed Size  : 337.63 MiB
> Packager        : Evangelos Foutras <foutrelis@archlinux.org>
> Build Date      : Fri 12 Nov 2021 06:01:59 PM CST
> Install Date    : Sun 14 Nov 2021 01:41:04 AM CST
> Install Reason  : Explicitly installed
> Install Script  : No
> Validated By    : Signature

**Comment by Sherlock-Holo** [comment]
I don't if the problem cause by rust toolchain or not, today I update the nightly toolchain to `rustc 1.58.0-nightly (b416e3892 2021-11-13)`, and rerun `cargo bpf build`, no error happened...

**Comment by rhdxmr** [comment]
@Sherlock-Holo 

Thank you for filing this issue.
I found that the Linux kernel v5.15 started to include `struct bpf_timer` in its source code. And also it is recorded in the kernel's BTF info. i.e., `/sys/kernel/btf/vmlinux`.

When RedBPF compiles redbpf-probes, it generates rust bindings for the data structures of the Linux kernel.
There are two different ways to find definitions of the Linux kernel data structures.
1. Using the Linux kernel header: `/lib/modules/$(uname -r)/`
2. Using the `/sys/kernel/btf/vmlinux`

RedBPF first tries finding the Linux kernel headers in the `/lib/modules/$(uname -r)`, but it fails to find the headers it tries to use `/sys/kernel/btf/vmlinux` as a fallback method. (See this for more information
https://docs.rs/redbpf-probes/2.1.0/redbpf_probes/#rules-about-method-selection)

I guess the error message you pasted in the text had occurred while redbpf used `vmlinux`.


Since the Linux kernel v5.15 contains `struct bpf_timer` and also [vmlinux_supplement.h](https://github.com/foniod/redbpf/blob/main/redbpf-probes/include/vmlinux_supplement.h) provides forward declaration of it, the error of multiple redefinition of `struct bpf_timer` had occurred.

Thanks to your report, I just got to know that Linux kernel v5.15 provides `struct bpf_timer`.
I'll fix this problem soon.

Thank you
